sdk/typescript/src/providers/json-rpc-provider.ts

```diff
@@ -80,29 +78,70 @@ export class JsonRpcProvider extends Provider {
    * the version compatibility of the SDK, as not all the schema
    * changes in the RPC response will affect the caller, but the caller needs to
    * understand that the data may not match the TypeSrcript definitions.
-   * @param rpcAPIVersion controls which type of RPC API version to use.
+   */
+  skipDataValidation?: boolean;
+  /**
+   * Configuration options for the websocket connection
+   */
+  socketOptions?: WebsocketClientOptions;
+  /**
+   * Cache timeout in seconds for the RPC API Version
+   */
+  versionCacheTimoutInSeconds?: number;
+};
+
+const DEFAULT_OPTIONS: RpcProviderOptions = {
+  skipDataValidation: true,
+  socketOptions: DEFAULT_CLIENT_OPTIONS,
+  versionCacheTimoutInSeconds: 600,
+};
+
+export class JsonRpcProvider extends Provider {
+  protected client: JsonRpcClient;
+  protected wsClient: WebsocketClient;
+  private rpcApiVersion: RpcApiVersion | undefined;
+  /**
+   * Establish a connection to a Sui RPC endpoint
+   *
+   * @param endpoint URL to the Sui RPC endpoint
+   * @param options configuration options for the provider
    */
   constructor(
     public endpoint: string,
-    public skipDataValidation: boolean = true,
-    // TODO: Update the default value after we deploy 0.12.0
-    private rpcAPIVersion: string = PRE_PAGINATION_API_VERSION,
-    public socketOptions: WebsocketClientOptions = DEFAULT_CLIENT_OPTIONS
+    public options: RpcProviderOptions = DEFAULT_OPTIONS
   ) {
     super();
 
+    const opts = { ...DEFAULT_OPTIONS, ...options };
+
     this.client = new JsonRpcClient(endpoint);
     this.wsClient = new WebsocketClient(
       endpoint,
-      skipDataValidation,
-      socketOptions
+      opts.skipDataValidation!,
+      opts.socketOptions
     );
   }
 
-  async getRpcApiVersion(): Promise<string> {
-    // TODO: we should fetch the API version from
-    // the RPC endpoint instead
-    return this.rpcAPIVersion;
+  async getRpcApiVersion(): Promise<RpcApiVersion | undefined> {
+    if (this.rpcApiVersion) {
+      return this.rpcApiVersion;
+    }
+    try {
+      const resp = await this.client.requestWithType(
+        'rpc.discover',
+        [],
+        isAny,
+        this.options.skipDataValidation
+      );
+      this.rpcApiVersion = parseVersionFromString(resp.info.version);
+      setTimeout(() => {
```

Comment:
Rather than holding a timeout, I think this would be better implemented as a timestamp that you can check against when you call this function. This would avoid holding an open timer. That would also make this easier to disable entirely if desired.

Reply:
done!